### ISSUE TITLE
ExecuteAsync question

### ISSUE THREAD

**Issue by bencassie** [roadrunnerlenny/etlbox#37]
Hi there,

Would it be possible to alter the ExecuteAsync method(s) to return their Task, rather than awaiting the Task?

I would like to be able to kick off multiple DBSources in parallel and then await them as needed.

Kind regards,
Ben

**Comment by roadrunnerlenny** [comment]
Hi Ben,

I just checked the implementation of ExecuteAsync() - https://github.com/roadrunnerlenny/etlbox/blob/c285a80dc9807c7c415c57ded86c3bba521c3fe7/ETLBox/src/Definitions/TaskBase/DataFlowSource.cs#L20

It does start a new task and is marked as async. The return Type is a Task. There is no explicit implementation which return the task - this is done by the C# compiler (or something else in the compiling process) - so the last await of the Task will implicitly return the Task.

So in the moment where you call ExecuteAsync, e.g. with 
Task t = source.ExecuteAsync()
this will fire a new Task starting the data flow without blocking your application. 
The task can then be used to check if the source is completed or not. 

So from my understanding you get a Task object here. Did I understand your question right?

**Comment by bencassie** [comment]
Hi there,

I wasn't aware that the last awaited task is automagically returned by the compiler - that's pretty cool! I suppose it does not make much sense to await the source anyhow.

My attention turns to the Destination.Wait() methods now and I see that it waits on its ActionBlock to complete before calling CleanUp(), making it problematic if you wanted to return the task for the caller to await.

I think it would be cool to be able to write something like this rather than individually calling Wait() on each destination:

```c#
DBDestination dest1 = new DBDestination("table1");
CSVDestination dest2 = new CSVDestination("dest2.csv");

Task.WaitAll(dest1.Wait(), dest2.Wait())
```
I guess in order to be able to return the awaitable task you would need to implement IDisposable on the destinations in order to facilitate the CleanUp() calls?

What do you think?

**Comment by roadrunnerlenny** [comment]
#28 #30

**Comment by roadrunnerlenny** [comment]
Hi Ben,

if you use the `ExecuteAsync()` method (which returns a task) you could use the `Wait()` on the destination, but I would not recommend it.

When I added the ExecuteAsync method, I also added an async method `Completion()` on all destination components. This will also return a task. 

So your code could look like this:
```C#
CustomSource source = new CustomSource(
        () => throw new ETLBoxException("Test Exception"), () => false
);
CustomDestination dest = new CustomDestination(row => {; });
source.LinkTo(dest);

Task sourceT = source.ExecuteAsync();  //Source task completes when all data is posted
Task destT = dest.Completion();   //Destination task completes when all data arrived at the source

Task.WhenAny(sourceT,destT);  //Be careful if you source throws an exception, this could not be handled
```

In this example, the source task completes when all data is posted into the data flow. The destination task completes when all data arrived at the source. 

See also issues #28 & #30. 

Does that help you?

**Comment by bencassie** [comment]
Ahh! I hadn't seen the Completion() method - many thanks that is perfect.
